### ISSUE TITLE
`aws sso login` should respect credential expiration

### ISSUE THREAD

**Issue by benkehoe** [aws/aws-cli#6948]
### Describe the feature

Currently, `aws sso login` goes through the log in process whenever it's called. So if I have two profiles `foo` and `bar` that share the same AWS SSO configuration (i.e., `sso_start_url` and `sso_region`), when I call `aws sso login --profile foo`, it will pop up a browser window, log me in, etc., resulting in a valid token being cached in `~/.aws/sso/cache`. If I then call `aws sso login --profile bar` (or `aws sso login --profile foo` again, for that matter), despite having a valid token for that config, it will go through the entire login process again, resulting in a new token.

### Use Case

It's unnecessary work, and it's exacerbated by the argument to `aws sso login` being `--profile`, which leads people to think they need to call it whenever they're using a particular profile (I wrote [an explainer](https://ben11kehoe.medium.com/you-only-need-to-call-aws-sso-login-once-for-all-your-profiles-41a334e1b37e) to help people understand they don't need to do that).

### Proposed Solution

`aws sso login` should check the expiration of the cached token, and exit out successfully if the token is not expired. There should be a `--force-refresh` option to override this.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CLI version used

N/A

### Environment details (OS name and version, etc.)

N/A

**Comment by lorengordon** [comment]
So much this, yes, please!

**Comment by gondalez** [comment]
I came here to report/request this also 🙏 💯 

My team and I use `aws sso login` during development and found having to log in breaks our concentration multiple times a day.

We use this fairly crude workaround script, but it is messy to make available to all of our usages.
I would much rather delete it and use `aws sso login` directly 😌 !

```bash
if aws sts get-caller-identity > /dev/null ; then
  echo '🔑 AWS Credentials AOK ✅'
else
  echo '🔑 AWS Credential NOK ❎. Logging in via SSO...'
  aws sso login
fi
```

**Comment by nmiekley** [comment]
Thanks for working on this! This would also solve #5972, wouldn't it?

**Comment by benkehoe** [comment]
I should note that `aws-sso-util login` does this already. https://github.com/benkehoe/aws-sso-util/blob/master/docs/login.md

**Comment by ixti** [comment]
@gondalez that's awesome. I've modified your solution for myself a bit:

``` bash
message="$(aws sts get-caller-identity 2>&1)"
exit_code="$?"

case "$exit_code" in
  0)
    echo -e "\e[32m🔑 AWS SSO credentials are valid.\e[0m"
    ;;

  255)
    echo -e "\e[33m🔑 AWS SSO credentials are expired. Logging in...\e[0m"
    aws sso login
    ;;

  *)
    echo -e "\e[31m🔑 Something went wrong...\e[0m" >&2
    echo "$message" >&2
    exit $exit_code
esac
```

**Comment by TheJosh** [comment]
One-liner that's also quiet:

```.bash
aws sts get-caller-identity >/dev/null 2>&1 || aws sso login
```

**Comment by kjenney** [comment]
Using `aws sts get-caller-identity` is NOT a foolproof solution. I just had a moment where `aws sts get-caller-identity` was returning a caller, but the session had expired. I still need a way to get the expiration timestamp to compare in case this happens again.

**Comment by benkehoe** [comment]
I mean, if you want to reach into the cache entry, this would work I think:

```bash
[ $(date -d "$(jq -r .expiresAt ~/.aws/sso/cache/<name>.json)" +"%s") -gt $(date +"%s") ] || aws sso login
```

with the (hashed) name in your particular setup there.

**Comment by kjenney** [comment]
I'm running BSD `date` on a Mac. I ended using this to get the expiration epoch - 

```
aws configure get sso_start_url --profile ${AWS_PROFILE} | xargs -I {} grep -h {} ~/.aws/sso/cache/*.json | jq .expiresAt | xargs -I {} sh -c 'TZ="UTC" date -j -f "%Y-%m-%dT%H:%M:%S" "+%s" $1 2>/dev/null' -- {} | head -n 1
```

Once I had that, I could compare the current epoch and followed with this:

```
# Where $1 is the expiration epoch
do_we_have_a_valid_session() {
        if [[ $(do_we_have_a_session) = "true" && $(has_our_session_expired $1) = "false" ]]; then echo "true"; else echo "false"; fi
}
```

**Comment by ixti** [comment]
Less convinient (in some cases) but pretty working solution is to use [aws-vault](https://github.com/99designs/aws-vault):

``` bash
aws-vault exec $profile -- aws s3 ls
```

Or:

``` bash
source <(aws-vault export $profile | grep '^AWS_' | sed 's/^/export /')
aws s3 ls
```

**Comment by ashovlin** [comment]
Hey all, revisiting this as we evaluate open feature requests. We are open to a contribution if somebody gets here before we do.

We don't want to change the default behavior of `aws sso login` for backwards compatibility. Rather than newly asking users to provide `--force` to match existing behavior, I'd rather gate the new behavior behind a new `--skip-if-valid` option (open to naming suggestions there).

I prefer to check if the access and refresh token validity, not whether the temporary credentials from `GetRoleCredentials` are still valid. As pointed out above: `aws sso login` is operating on the `sso-session` as opposed to a specific profile/role, and the `~/.aws/cli/cache` directory can muddy the waters.

So for `--skip-if-valid`, I would:
1. Locate the right token cache file
2. If the access token's `expiresAt` is still valid, return success.
3. If the refresh token can successfully still be exchanged, do so, and return success.
4. Otherwise proceed with the full, current workflow.

**Comment by gunzy83** [comment]
> Hey all, revisiting this as we evaluate open feature requests. We are open to a contribution if somebody gets here before we do.
> 
> We don't want to change the default behavior of `aws sso login` for backwards compatibility. Rather than newly asking users to provide `--force` to match existing behavior, I'd rather gate the new behavior behind a new `--skip-if-valid` option (open to naming suggestions there).
> 
> I prefer to check if the access and refresh token validity, not whether the temporary credentials from `GetRoleCredentials` are still valid. As pointed out above: `aws sso login` is operating on the `sso-session` as opposed to a specific profile/role, and the `~/.aws/cli/cache` directory can muddy the waters.
> 
> So for `--skip-if-valid`, I would:
> 
> 1. Locate the right token cache file
> 2. If the access token's `expiresAt` is still valid, return success.
> 3. If the refresh token can successfully still be exchanged, do so, and return success.
> 4. Otherwise proceed with the full, current workflow.

This looks great to me. Most of us are just looking for a way to automatically check the validity of the session rather than any specific role that can be assumed from that session and login again if it is not valid. Doing this in one command just makes our existing scripts/taskfiles etc more simple and reliable.

**Comment by jorhett** [comment]
A command to just return 0 or 1 for is-valid (without a login) would also be useful.  So steps 1-3 without step 4?

**Comment by benkehoe** [comment]
> 1. Locate the right token cache file
> 2. If the access token's expiresAt is still valid, return success.
> 3. If the refresh token can successfully still be exchanged, do so, and return success.
> 4. Otherwise proceed with the full, current workflow.

Agreed on this being the right process, but I'd like to push back a little bit on `--skip-if-valid`, which requires that people learn about it. If we went the `--force-refresh` route, so executing the above process by default, and if a valid token was found or retrieved the CLI printed:

> Identity Center session appears to still be valid and login not needed. To force a fresh login, use --force-refresh.

what are the use cases that we think would break? I would argue the vast majority of users would be happy not to have to go through the login flow and their use cases would not break. The ones where something still wasn't working, they would get instructions for it directly.

Maybe an alternative is to run the check every time anyway, and if a valid token is found, print "You appear to have a valid session. You can use --skip-if-valid to avoid a fresh login when it is not needed." before still proceeding with the current login flow. But I still think it's worth thinking hard if we can make skip-if-valid behavior the default.
